// limitReptile/index.js
"use strict";

// 依赖
const http = require('http');
const https = require('https');
const limit = require('limit');
const cheerio = require('cheerio');

const reptile = {};
const MySql = limit.createMySql({
	database:'thetop'
});

function mockHeader(header){
	return limit.assign({
		'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36',
	}, header);
};

// 虚拟请求头
function setConfig(config){
	let newConfig = limit.assign({
		protocol: 'http:',
		host: 'localhost',
		path: '/',
		method: 'GET',
		headers: {}
	}, config);
	if( newConfig.protocol === 'http:' ){
		newConfig.port = '80';
	}else{
		newConfig.port = '443';
	};
	return newConfig;
};

// 解析文档
/**
 * 数据结构
 * [{
 * 		name: '毕业会考',
 * 		score: 7.9,
 * 		director: '克里斯蒂安·蒙吉',
 * 		actor: '阿德里安·蒂蒂耶尼,玛丽亚·维多利亚·德拉格斯,拉雷什·安德瑞斯',
 * 		area: '英国',
 * 		type: '爱情',
 * 		year: 2016
 * }]
 */
function parseHtml(data){
	let $ = cheerio.load(data);
	let nameList = $('.doulist-item').map((key, val) => {
		val = $(val);
		let name = val.find('.title a').text().trim().split(' ')[0];
		let score = +val.find('.rating_nums').text().trim();
		let abstract = parseAbstract( val.find('.abstract').text().trim() );
		let director = abstract['主演']||'';
		let actor = abstract['导演']||'';
		let type = abstract['类型']||'';
		let area = abstract['制片国家/地区']||'';
		let year = abstract['年份']||'';
		let src = val.find('.post img').prop('src');
		return {name, score, director, actor, area, type, year, src};
	});
	nameList = Array.from(nameList);
	// return console.log(nameList);
	return MySql.insertData(nameList);
};

function parseAbstract(str){
	let map = {};
	let list = str.split(/\n/).filter((val, key) => {
		val = val.trim();
		if( val.trim() ){
			return true;
		};
	}).map((val, key) => {
		return val.trim();
	}).forEach((v) => {
		let str = v.split(':');
		let key = str[0];
		let val = splitStr(str[1]);
		map[key] = val;
	});
	return map;
};

function splitStr(str){
	return str.split('/').map((val) => {
		return val.trim();
	}).join(',')
};

// '/doulist/43710634/?start=125&sort=time&sub_type=1'
function get(list){
	let proList = list.map((val) => {
		return new Promise((s, j) => {
			let config = setConfig({
				protocol: 'https:',
				host: 'www.douban.com',
				path: val,
				headers: mockHeader()
			});
			let httpMethod = config.protocol === 'http:' ? http : https;
			let req = httpMethod.request(config, (res) => {
				let statusCode = res.statusCode;
				// res.setEncoding('utf8');
				let rawData = '';
				res.on('data', (chunk) => rawData += chunk);
				res.on('end', () => {
					s( parseHtml(rawData) );
			  	});
			})
			req.end();
		});
	});
	Promise.all(proList).then(() => {
		MySql.end();
	});
};

function insertMovie(){
	let list = [];
	// 2016年电影数据
	// for(let i = 0; i < 8; i++){
	// 	list.push(`/doulist/43710634/?start=${i*25}&sort=time&sub_type=1`);
	// };
	// 2015年电影数据
	for(let i = 0; i < 15; i++){
		list.push(`/doulist/36742433/?start=${i*25}&sort=time&sub_type=1`);
	};
	get(list);
};

function insertSort(){
	let list = [{ name: '吉星高照2015',
	    score: 3.1,
	    director: '曾志伟,王祖蓝,陈嘉桦',
	    actor: '晴朗',
	    area: '香港,中国大陆,台湾',
	    type: '喜剧',
	    year: '2015',
	    src: 'https://img3.doubanio.com/view/movie_poster_cover/lpst/public/p2229854022.jpg' },
	  { name: '超能陆战队',
	    score: 8.6,
	    director: '斯科特·安第斯,瑞恩·波特,丹尼尔·海尼',
	    actor: '唐·霍尔,克里斯·威廉姆斯',
	    area: '美国',
	    type: '喜剧,动作,科幻,动画,冒险',
	    year: '2014',
	    src: 'https://img1.doubanio.com/view/movie_poster_cover/lpst/public/p2224568669.jpg' }];
	
};

reptile.start = function(){
	// 插数据
	// insertMovie();
	// 插类型
	insertSort();
};

module.exports = reptile;
