// limitReptile/movie.js
"use strict";

// 依赖
const http = require('http');
const https = require('https');
const limit = require('limit');
const cheerio = require('cheerio');

const reptile = {};
const MySql = limit.createMySql({
	database:'thetop'
});

function mockHeader(header){
	return limit.assign({
		'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.116 Safari/537.36',
	}, header);
};

// 虚拟请求头
function setConfig(config){
	let newConfig = limit.assign({
		protocol: 'http:',
		host: 'localhost',
		path: '/',
		method: 'GET',
		headers: {}
	}, config);
	if( newConfig.protocol === 'http:' ){
		newConfig.port = '80';
	}else{
		newConfig.port = '443';
	};
	return newConfig;
};

// 解析文档
/**
 * 数据结构
 * [{
 * 		name: '毕业会考',
 * 		score: 7.9,
 * 		director: '克里斯蒂安·蒙吉',
 * 		actor: '阿德里安·蒂蒂耶尼,玛丽亚·维多利亚·德拉格斯,拉雷什·安德瑞斯',
 * 		area: '英国',
 * 		type: '爱情',
 * 		year: 2016
 * }]
 */
function parseHtml(data){
	let $ = cheerio.load(data);
	let nameList = $('.doulist-item').map((key, val) => {
		val = $(val);
		let name = val.find('.title a').text().trim().split(' ')[0];
		let score = +val.find('.rating_nums').text().trim();
		let abstract = parseAbstract( val.find('.abstract').text().trim() );
		let director = abstract['主演']||'';
		let actor = abstract['导演']||'';
		let type = abstract['类型']||'';
		let area = abstract['制片国家/地区']||'';
		let year = abstract['年份']||'';
		let src = val.find('.post img').prop('src');
		return {name, score, director, actor, area, type, year, src};
	});
	nameList = Array.from(nameList);
	// return console.log(nameList);
	return MySql.insertData(nameList);
};

function parseAbstract(str){
	let map = {};
	let list = str.split(/\n/).filter((val, key) => {
		val = val.trim();
		if( val.trim() ){
			return true;
		};
	}).map((val, key) => {
		return val.trim();
	}).forEach((v) => {
		let str = v.split(':');
		let key = str[0];
		let val = splitStr(str[1]);
		map[key] = val;
	});
	return map;
};

function splitStr(str){
	return str.split('/').map((val) => {
		return val.trim();
	}).join(',')
};

// '/doulist/43710634/?start=125&sort=time&sub_type=1'
function get(list){
	let proList = list.map((val) => {
		return new Promise((s, j) => {
			let config = setConfig({
				protocol: 'https:',
				host: 'www.douban.com',
				path: val,
				headers: mockHeader()
			});
			let httpMethod = config.protocol === 'http:' ? http : https;
			let req = httpMethod.request(config, (res) => {
				let statusCode = res.statusCode;
				// res.setEncoding('utf8');
				let rawData = '';
				res.on('data', (chunk) => rawData += chunk);
				res.on('end', () => {
					s( parseHtml(rawData) );
			  	});
			})
			req.end();
		});
	});
	Promise.all(proList).then(() => {
		MySql.end();
	});
};

function insertMovie(){
	let list = [];
	// 2016年电影数据
	// for(let i = 0; i < 8; i++){
	// 	list.push(`/doulist/43710634/?start=${i*25}&sort=time&sub_type=1`);
	// };
	// 2015年电影数据
	for(let i = 0; i < 15; i++){
		list.push(`/doulist/36742433/?start=${i*25}&sort=time&sub_type=1`);
	};
	get(list);
};

function insertSort(){
	MySql.query('select * from movie').then((val) => {
		let proList = val.map((val) => {
			let mid = val.id;
			if( val.name ){
				return MySql.query(`select * from sort where mid=${mid}`).then((midList) => {
					if( !midList.length ){
						let arr = limit.flatten([
							insertData(val.year, 2, mid),
							insertData(val.type, 3, mid),
							insertData(val.area, 4, mid)
						]);
						return Promise.all(arr);
					};
				});
			};
		});
		Promise.all(proList).then(() => {
			MySql.end();
		}).catch((e) => {limit.err(e)});
	});
};

function insertData(key, pid, mid){
	return key.split(',').map((name) => {
		// console.log({name, pid, mid});
		return MySql.insertData('sort', [{name, pid, mid}]);
	});
};

function resetData(){
	MySql.query('DELETE FROM sort where id > 4;').then(() => {
		return MySql.query('ALTER table sort AUTO_INCREMENT = 5;');
	}).catch((e) => {limit.err(e)}).finally(() => {
		MySql.end();
	});
};

reptile.start = function(){
	// 插数据
	// insertMovie();
	// 插类型
	insertSort();
	// 清楚重置
	// resetData();
};

module.exports = reptile;
