// limit/mysql.js
"use strict";

// 依赖
const mysql = require('mysql');
const rootPath = __dirname.split('node_modules')[0];

class Connection {
	constructor(props){
		let me = this;
		let state = me.state = {
			host: 'localhost',
		    user: 'root',
		    password: '',
		    database:'',
		    port: 3306
		};
		Object.assign(state, props);
		me.pool = mysql.createPool(state);
		me.getMainSql();
	}
	getMainSql(){
		let me = this;
		me.mainSql = require(`${rootPath}/sql/${me.state.database}`);
	}
	querySuper(queryName, ...args){
		let me = this;
		let {mainSql} = me;
		let fnSql = mainSql[queryName];
		if( typeof fnSql === 'function' ){
			return me.query( fnSql(...args) );
		}else{
			return me.query( ''+fnSql );
		};
	}
	query(query){
		let me = this;
		let pool = me.pool;
		return new Promise((s, j) => {
			pool.getConnection((err, conn) => {
			    if( err ){
			    	j(err);
			    }else{
			    	conn.query(query, (qerr, ...args) => {
			    		conn.release();
			    		if( qerr ){
			    			j(qerr)
			    		}else{
			    			s(...args);
			    		};
			    	});
			    };
			});
		});
	}
	insertData(list){
		let me = this;
		let {state} = me;
		return new Promise((s, j) => {
			if( list.length ){
				let keys = Object.keys(list[0]).join(',');
				let vals = list.map( (val) => {
					return Object.values(val).map((val) => {
						return JSON.stringify(val);
					}).join(',');
				} ).map((val) => {
					return `(${val})`;
				}).join(',');
				me.query(`insert into ${state.database} (${keys}) values ${vals}`).then(s, j);
			}else{
				j('list is empty');
			};
		});
	}
	end(){
		let me = this;
		me.pool.end();
	}
};

module.exports = Connection;
